### ISSUE TITLE
Add Support for Broader Set of Illumina Beadchip Platforms

### ISSUE THREAD

**Issue by Miserlou** [AlexsLemonade/refinebio#232]
### Context

Quoting @jaclyn-taroni, via https://github.com/AlexsLemonade/refinebio/pull/212#discussion_r184864928:

```
This only supports 2 human platforms, Illumina HumanWG-6 v2.0 (i.e., GPL13376) and Illumina HumanHT-12 V4.0 (i.e., GPL10558), using the relevant bioconductor packages at the moment.

We'll want to support a broader set of BeadChips. Specifically, there are bioconductor packages for the following additional platforms:

    illuminaHumanv1.db
    illuminaHumanv3.db
    illuminaMousev1.db
    illuminaMousev1p1.db
    illuminaMousev2.db
    illuminaRatv1.db

You also might consider getting the probe objects like so:

probeSequenceRef <- lazyeval::lazy_eval(paste0(platform, ".db::", platform, "PROBESEQUENCE"))
probeQualityRef <- lazyeval::lazy_eval(paste0(platform, ".db::", platform, "PROBEQUALITY"))
probeGeneRef <-  lazyeval::lazy_eval(paste0(platform, ".db::", platform, "ENSEMBLREANNOTATED"))

☝️ would also eliminate the need for loading the annotation library (line 356).
```

### Problem or idea

Until we explicitly add support for other platforms, we'll only be able to process results from two human Illumina platforms. 

### Solution or next step

Jackie has also ready described the solution, so I think it just needs to be implemented. There may be a little bit of work to detect the platform as well.

### New Issue Checklist

- [x] The title is short and descriptive.
- [x] You have explained the context that led you to write this issue.
- [x] You have reported a problem or idea.
- [x] You have proposed a solution or next step.

**Comment by Miserlou** [comment]
This ticket would be greatly enhanced by a small suite of example accessions.

**Comment by jaclyn-taroni** [comment]
Will do today while wrapping up the microarray whitelist

**Comment by jaclyn-taroni** [comment]
While working on #222/#228 today, I've found that relationship between Illumina (bead array) platforms (and therefore their annotation) is less straightforward than I initially imagined. Specifically, I thought we would just support the platforms that the Bioconductor packages are named for and that was that. However, there are a considerable amount of samples we'd miss out on that way. 

Here's a `v3.0` example:

[`GPL6884`](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6884), Illumina HumanWG-6 v3.0 expression beadchip, has over 8000 samples.  [`GPL6883`](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6883), Illumina HumanRef-8 v3.0 expression beadchip, has almost 7000 samples.

The description for [`illuminaHumanv3.db`](https://doi.org/doi:10.18129/B9.bioc.illuminaHumanv3.db) is 
> Illumina HumanHT12v3 annotation data (chip illuminaHumanv3) assembled using data from public repositories

This corresponds to [`GPL6947`](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6947).

However, the accepted answer to [this Bioconductor question](https://support.bioconductor.org/p/62661/) notes that the HT-12 (`GPL6947`) and WG-6 (`GPL6884`) arrays have the same content. The [Gemma listing for Ref-8 v3.0 (`GPL6883`)](https://gemma.msl.ubc.ca/arrays/showArrayDesign.html?id=376) lists this platform as being subsumed by `GPL6947`. (That is not indicated in [the listing for WG-6/`GPL6884`](https://gemma.msl.ubc.ca/arrays/showArrayDesign.html?id=443).)

When I check the GPL info from GEO, I find that all the `ID` from `GPL6883` and `GPL6884` are contained within the `ID` from `GPL6947`. 

Based on the above Bioconductor answer + what I've laid out + the number of IDs on each of these platforms, I think the relationship between WG-6 and HT-12 is the number of samples that are supported and my suspicion is that the Ref-8 was an earlier product.

Here's the GPL info from GEO for those platforms: [Illumina_v3.0_GPL.zip](https://github.com/AlexsLemonade/refinebio/files/2006060/Illumina_v3.0_GPL.zip)

Anyway, I want to do a bit more research and this will take me a bit longer to normalize/whitelist than I expected.

**Comment by jaclyn-taroni** [comment]
Alright, I think the WG arrays are just Ref arrays with probes beyond RefSeq content. From [this Illumina documentation](https://support.illumina.com/content/dam/illumina-marketing/documents/products/datasheets/datasheet_gene_exp_analysis.pdf) (_Array-Based Gene Expression Analysis_):

> The MouseWG-6 v2.0 and MouseRef-8 v2.0 Expression BeadChips are Illumina’s genome-wide gene expression profiling solutions for researchers using the mouse as their model organism. The MouseRef-8 BeadChip features recent content derived from the NCBI RefSeq database. In addition to RefSeq content, the MouseWG-6 BeadChip
is supplemented with probes that target the Mouse Exonic Evidence Based Oligonucleotide (MEEBO) set2, as well as exemplar protein- coding sequences described in the RIKEN FANTOM24-6 database

**Comment by jaclyn-taroni** [comment]
I've just gone through the human platforms. Here's a Bioconductor package to platform name mapping that I think should work:

`illuminaHumanv1.db`: Illumina Human-6 v1.0, Illumina HumanRef-8 v1.0
`illuminaHumanv2.db`: Illumina Human-6 V2.0, Illumina HumanRef-8 v2.0
`illuminaHumanv3.db`: Illumina HumanHT-12 V3.0, Illumina HumanRef-8 v3.0, Illumina HumanWG-6 V3.0
`illuminaHumanv4.db`: Illumina HumanHT-12 V4.0

I'm thinking we might want to be a bit more stringent here:
https://github.com/AlexsLemonade/refinebio/blob/dev/workers/data_refinery_workers/processors/illumina.R#L414

Specifically, we might want some (high) percentage of the expression data probes and the probes from the packages to overlap and for the processing to fail if they don't (i.e., move that check to before the SCAN norm parts that are more intensive).

**Comment by jaclyn-taroni** [comment]
Bioconductor package mapping for mouse and rat:

`illuminaMousev1.db`: Illumina Mouse-6 V1.0, Illumina MouseRef-8 V1.0
`illuminaMousev1p1.db`: Illumina Mouse-6 v1.1, Illumina MouseRef-8 V1.1
`illuminaMousev2.db`: Illumina MouseWG-6 v2.0, Illumina MouseRef-8 v2.0
`illuminaRatv1.db`: Illumina RatRef-12 V1.0

**Comment by jaclyn-taroni** [comment]
Alright, so first step to trying to distinguish platforms based on their IDs is to check for overlap _between_  platforms. I'm posting about human whole genome chips here. I'm using the GEO accession that has the most samples assigned to it.

| Platform Name | GEO Accession |
|----------------|-----------------|
| Human-6 v1.0 | [GPL2507](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL2507) |
| Human-6 v2.0 | [GPL6102](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6102) |
| HumanHT-12 v3.0 | [GPL6947](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6947) |
| HumanHT-12 v4.0 | [GPL10558](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL10558) |

Here's a Venn diagram of the overlap (using the `ID` column):

![test](https://user-images.githubusercontent.com/19534205/40178378-54ecca2e-59af-11e8-894e-94e170704d9d.png)

We can see that checking to make sure something is version 1.0 should be pretty straightforward. The other platforms share some proportion of their IDs.

There is [functionality in the `beadarray` package](https://www.rdocumentation.org/packages/beadarray/versions/2.22.2/topics/Annotation) that will suggest the annotation based on the platform with the highest overlap with the input data. (This will require [the use of `beadarray` functions to get a `beadLevelData-class` object](https://www.rdocumentation.org/packages/beadarray/versions/2.22.2/topics/readIllumina).)

For human and mouse, distinguishing between "WG" and "Ref" should be straightforward -- the "Ref" platforms measure about half the number of transcripts.

**Comment by jaclyn-taroni** [comment]
I have some quick analyses for "platform detection here: https://github.com/jaclyn-taroni/beadarray-platform-detection/tree/61bc2700e34b07f78672e32a9aa3544f150a4718

I will briefly summarize what's laid out in the README here for convenience.

#### Methods overview

* Obtained a list of human Illumina probes from  `v1.26.0` of the following packages: [`illuminaHumanv1.db`](https://bioconductor.org/packages/illuminaHumanv1.db/), [`illuminaHumanv2.db`](https://bioconductor.org/packages/illuminaHumanv2.db/), [`illuminaHumanv3.db`](https://bioconductor.org/packages/illuminaHumanv3.db/), [`illuminaHumanv4.db`](https://bioconductor.org/packages/illuminaHumanv4.db/)
* Selected 30 random GEO series accessions (experiments) each from `GPL2507`, `GPL6102`, `GPL6947`, and `GPL10558`. I restricted the experiments to those that  had supplementary TXT files & downloaded any supp. files matching this pattern `*non*`.
* Calculated the overlap between whatever was in the _first column_ of the non-normalized files (termed "IDs") and the 4 probes lists. _(This is admittedly not a great strategy, but it is useful to see how often the `ID_REF` as the first column that contains probes assumption holds.)_

I was interested in both the proportion of IDs  found in the probes AND the proportion of probes found in the IDs. This is because the Ref platforms typically have somewhere around half the probes from the WG platforms.

#### Results

I expected to find that the IDs in the non-normalized file for the majority of series/experiments. (As a reminder, `GPL2507`::`Humanv1`, `GPL6102`::`Humanv2`, `GPL6947`::`Humanv3`, and `GPL10558`::`Humanv4`.)

This is what I observed. Here's a heatmap showing the relationship between platforms (e.g., rows, `GPLXXXX`) and the probe lists from the Bioconductor packages (e.g., columns, `HumanvX`) . The cells represent the median proportion of the IDs from the experiments from that platform that were found in 
I'll note that this is follows a pattern I was expecting based on the venn diagram I posted earlier.

![](https://github.com/jaclyn-taroni/beadarray-platform-detection/blob/61bc2700e34b07f78672e32a9aa3544f150a4718/plots/ids_heatmap.png?raw=true)

For the most part, a high percentage of the probes were in the IDs as well. 

There were, of course, series that deviated from this expected pattern. I go into the `GPL6947` examples [here](https://github.com/jaclyn-taroni/beadarray-platform-detection#results).

They were typically the result of mislabeling (on my part, not on the part of the submitters; I didn't take into account that samples within an experiment could be from multiple platforms.), use of different identifiers (e.g., gene symbols or bead IDs rather than probe IDs), or the probe IDs were in a different column (e.g., not the _first_ one). 

I'll note that in the one example of a Ref chip, less than half the probes were present in the IDs.

**Comment by jaclyn-taroni** [comment]
Our pipeline might look like this for human and mouse* (after first obtaining a list of probes from the _most recent version_ of the bioconductor packages):

* Calculate the proportion of IDs (using whatever method we're using to identify the ID column currently) in each of the probe lists and the proportion of probes in the IDs. Store this information (8 numbers in the human case) for each experiment. (Done in `illumina.py`?)
* Suggest "WG" vs. "Ref" based on % of probes in IDs (more work needed here to pick a threshold!!).
* Use the package with the highest number of ids in probes list if it is >95% and pass this as an argument to `illumina.R` (also store it as some kind of computational annotation). Else, fail.

We could use the same strategy where `platform` is the aforementioned argument to `illumina.R`

```R
probeSequenceRef <- lazyeval::lazy_eval(paste0(platform, ".db::", platform, "PROBESEQUENCE"))
probeQualityRef <- lazyeval::lazy_eval(paste0(platform, ".db::", platform, "PROBEQUALITY"))
probeGeneRef <-  lazyeval::lazy_eval(paste0(platform, ".db::", platform, "ENSEMBLREANNOTATED"))
```
*In rat, we'd still probably like to use the 95% threshold, but we only have one platform available.

**Next steps:**

I would suggest implementing the first part (calculating proportions & storing them) and trying some random (10s of?) series from more platforms because we've only looked at the most popular or official versions of the WG platforms.

Here are the additional GEO platforms I would suggest looking at: `GPL10295`, `GPL6106`, `GPL6104`, `GPL6255`, `GPL5104`, `GPL6098`, `GPL6883`, `GPL6884`, `GPL4234`, `GPL6333`, `GPL6105`, `GPL6481`, `GPL6422`, `GPL8389`, `GPL6103`, `GPL6885`, `GPL19575`, `GPL6887`, `GPL17543`, `GPL6101`, `GPL17223`

**Comment by cgreene** [comment]
Wow! I am loving the **Methods** and **Results** discussion above. That is a great way to make sure that we're accurate!

**Comment by jaclyn-taroni** [comment]
Just chatted with @Miserlou and here is some pseudocode for what we came up with for platform detection:
```
get & read in non-normalized supplementary data
IDs = identifiers used by submitter 

for each version of db package for organism (e.g., v1-v4 in human)
  calculate the % of IDs in the probes for current package
  calculate the % of probes in the IDs for current package

save the calculated overlaps as SampleAnnotation (will be number of versions x 2)
highest_overlap = which version has the highest % of IDs in the probes

if highest_overlap > 75%
  supply as platform/package to Illumina SCAN Rscript
else
  we should NOOP the processed data
```

The `% probes in the IDs` will help us distinguish between whole genome and ref chips, whereas the `% IDs in the probes` should be used for deciding whether or not to continue to process the non-normalized data. Recall that probe sequence information will be required to perform the normalization, so if many of the IDs cannot be mapped to probes and, therefore, their sequences, we will not be able to normalize them effectively.

**Comment by jaclyn-taroni** [comment]
To clarify `calculate the % of IDs in the probes` means `find the proportion or percentage of IDs that can be mapped to probes`

**Comment by jaclyn-taroni** [comment]
Closed by #340
